examples/verified-fetch/fetcher.go

```diff
@@ -0,0 +1,271 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strings"
+
+	"github.com/ipfs/boxo/blockservice"
+	"github.com/ipfs/boxo/blockstore"
+	"github.com/ipfs/boxo/exchange/offline"
+	bsfetcher "github.com/ipfs/boxo/fetcher/impl/blockservice"
+	files "github.com/ipfs/boxo/files"
+	"github.com/ipfs/boxo/ipld/merkledag"
+	unixfile "github.com/ipfs/boxo/ipld/unixfs/file"
+	"github.com/ipfs/boxo/ipns"
+	"github.com/ipfs/boxo/namesys"
+	"github.com/ipfs/boxo/path"
+	"github.com/ipfs/boxo/path/resolver"
+	"github.com/ipfs/go-datastore"
+	dssync "github.com/ipfs/go-datastore/sync"
+	"github.com/ipfs/go-unixfsnode"
+	gocarv2 "github.com/ipld/go-car/v2"
+	dagpb "github.com/ipld/go-codec-dagpb"
+	madns "github.com/multiformats/go-multiaddr-dns"
+)
+
+// fetcher fetches files over HTTP using verifiable CAR archives.
+type fetcher struct {
+	gateway   string
+	limit     int64
+	userAgent string
+}
+
+type fetcherOption func(f *fetcher) error
+
+// withUserAgent sets the user agent for the [Fetcher].
+func withUserAgent(userAgent string) fetcherOption {
+	return func(f *fetcher) error {
+		f.userAgent = userAgent
+		return nil
+	}
+}
+
+// withLimit sets the limit for the [Fetcher].
+func withLimit(limit int64) fetcherOption {
+	return func(f *fetcher) error {
+		f.limit = limit
+		return nil
+	}
+}
+
+// newFetcher creates a new [Fetcher]. Setting the gateway is mandatory.
+func newFetcher(gateway string, options ...fetcherOption) (*fetcher, error) {
```

Comment:
What's the reason we're doing this as a test rather than a library component? e.g. we don't have any consumers immediately setup to use it, not enough time to productionize it yet, trying to show people the guts more than give them the tools, etc.

Reply:
@aschmahmann well, this isn't a test. It's an example. But if we want to make it a part of Boxo itself, then we need to improve it massively and allow it to work with UnixFS directories too and not only files.